dali/pipeline/operators/decoder/nvjpeg_decoder_new.h

```diff
@@ -0,0 +1,268 @@
+
+  ~nvJPEGDecoderNew() noexcept(false) {
+    for (auto &j_s : jpeg_streams_) {
+      NVJPEG_CALL(nvjpegJpegStreamDestroy(j_s));
+    }
+    NVJPEG_CALL(nvjpegDecoderDestroy(decoder_huff_host_));
+    NVJPEG_CALL(nvjpegDecoderDestroy(decoder_huff_hybrid_));
+    NVJPEG_CALL(nvjpegDestroy(handle_));
+  }
+
+  using dali::OperatorBase::Run;
+  void Run(MixedWorkspace *ws) override {
+    /******************************************
+     ****************** CPU *******************
+    ******************************************/
+
+    for (int i = 0; i < batch_size_; i++) {
+      const auto &in = ws->Input<CPUBackend>(0, i);
+      const auto *input_data = in.data<uint8_t>();
+      const auto in_size = in.size();
+      // Get necessary image information
+
+      nvjpegStatus_t ret = nvjpegJpegStreamParse(handle_,
+                                                 static_cast<const unsigned char*>(input_data),
+                                                 in_size,
+                                                 1,
+                                                 1,
+                                                 jpeg_streams_[i]);
+
+      if (ret == NVJPEG_STATUS_BAD_JPEG) {
+        auto file_name = in.GetSourceInfo();
+        try {
+          EncodedImageInfo info;
+          const auto image = ImageFactory::CreateImage(
+                                            static_cast<const uint8 *>(input_data), in_size);
+          const auto dims = image->GetImageDims();
+          info.heights[0] = std::get<0>(dims);
+          info.widths[0] = std::get<1>(dims);
+          info.nvjpeg_support = false;
+          output_info_[i] = info;
+        } catch (const std::runtime_error &e) {
+          DALI_FAIL(e.what() + "File: " + file_name);
+        }
+      } else {
+        EncodedImageInfo info;
+        NVJPEG_CALL(nvjpegJpegStreamGetFrameDimensions(jpeg_streams_[i],
+                                                       info.widths,
+                                                       info.heights));
+        NVJPEG_CALL(nvjpegJpegStreamGetComponentsNum(jpeg_streams_[i],
+                                                           &info.c));
+        NVJPEG_CALL(nvjpegJpegStreamGetChromaSubsampling(jpeg_streams_[i],
+                                                         &info.subsampling));
+        info.nvjpeg_support = SupportedSubsampling(info.subsampling);
+        output_info_[i] = info;
+
+        if (ShouldBeHybrid(info)) {
+          image_decoders_[i] = decoder_huff_hybrid_;
+        } else {
+          image_decoders_[i] = decoder_huff_host_;
+        }
+
+        /*
+        // TODO(spanev): add this function when integrating Crop
+        nvjpegnvjpegDecodeParamsSetROI(decode_params_[pos], offset_x, offset_y, roi_w, roi_h);
+        */
+        NVJPEG_CALL(nvjpegStateAttachPinnedBuffer(image_states_[i],
+                                                  pinned_buffer_[i]));
+        NVJPEG_CALL(nvjpegDecodeJpegHost(
```

Comment:
You need to check status here and do the fallback if host part fail. We do that for failures in `nvjpegDecodePhaseOne`.

Reply:
Done